File: 2020/cpp/day4/day4_1.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <algorithm>

using namespace std;

class PassportCheck {
    private:
        int validCounter = 0;
        ifstream infile;

        // byr, iyr, eyr, hgt, hcl, ecl, pid
        bool arr[7] = {false};

    
    public:
        PassportCheck(string filename) {
            infile.open(filename);
        }

        void checkLine(string line) {
            if (line.find("byr") != -1) {
                arr[0] = 1;
            }
            if (line.find("iyr") != -1)
            {
                arr[1] = 1;
            }
            if (line.find("eyr") != -1)
            {
                arr[2] = 1;
            }
            if (line.find("hgt") != -1)
            {
                arr[3] = 1;
            }
            if (line.find("hcl") != -1)
            {
                arr[4] = 1;
            }
            if (line.find("ecl") != -1)
            {
                arr[5] = 1;
            }
            if (line.find("pid") != -1)
            {
                arr[6] = 1;
            }
        }

        void resetArr() {
            for (int i = 0; i < 7; i++) {
                arr[i] = 0;
            }
        }

        void checkEntry () {
            if (all_of(begin(arr), end(arr), [](bool i){return i;})) {
                validCounter++;
            }
        }

        int checkFile () {
            for (string line; getline(infile, line);) {
                if (line == "") {
                    checkEntry();
                    resetArr();
                } else {
                    checkLine(line);
                }

            }

            return validCounter;
        }
        
};

int main()
{
    PassportCheck machine("input.txt");

    cout << machine.checkFile() << endl;
}

File: 2020/python/day4/day4_1.py
#!/usr/bin/env python3
# Author: Mathias Rahbek-Borre
# Advent of code - day 4 part 1:
import re


infile = open("input.txt", "r")

valid_fields = 0
valid_passports = 0
for line in infile:
    byr = re.search(r"byr:", line)
    if byr:
        valid_fields += 1

    iyr = re.search(r"iyr:", line)
    if iyr:
        valid_fields += 1

    eyr = re.search(r"eyr:", line)
    if eyr:
        valid_fields += 1

    hgt = re.search(r"hgt:", line)
    if hgt:
        valid_fields += 1

    hcl = re.search(r"hcl:", line)
    if hcl:
        valid_fields += 1

    ecl = re.search(r"ecl:", line)
    if ecl:
        valid_fields += 1

    pid = re.search(r"pid:", line)
    if pid:
        valid_fields += 1

    # cid = re.search(r"cid:", line)
    # if cid:
    #     valid_fields += 1

    if line.strip() == "":
        if valid_fields == 7:
            valid_passports += 1
        valid_fields = 0

if valid_fields != 0:
    if valid_fields == 7:
        valid_passports += 1

print(valid_passports)
